daemon/access.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/polkit"
+)
+
+type accessResult int
+
+const (
+	accessOK accessResult = iota
+	accessUnauthorized
+	accessForbidden
+	accessCancelled
+)
+
+var (
+	polkitCheckAuthorization = polkit.CheckAuthorization
+	checkPolkitAction        = checkPolkitActionImpl
+)
+
+func checkPolkitActionImpl(r *http.Request, ucred *ucrednet, action string) accessResult {
+	var flags polkit.CheckFlags
+	allowHeader := r.Header.Get(client.AllowInteractionHeader)
+	if allowHeader != "" {
+		if allow, err := strconv.ParseBool(allowHeader); err != nil {
+			logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+		} else if allow {
+			flags |= polkit.CheckAllowInteraction
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	switch authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err {
+	case nil:
+		if authorized {
+			// polkit says user is authorised
+			return accessOK
+		}
+	case polkit.ErrDismissed:
+		return accessCancelled
+	default:
+		logger.Noticef("polkit error: %s", err)
+	}
+	return accessUnauthorized
+}
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+// openAccess allows requests without authentication, provided they
```

Comment:
There is some argument to made to call this "unrestricted" instead of "open", and then change the docs (which also atm aren't clear about the access restrictions on /v2/snapctl either) to make clear the role of the unix socket in the access. It seems something though to do in a self-contained follow up.